commit fe3fcffacd3380a23dcd3da3142a2fbfc403566a (refs/changes/28/2828/4)
Author: Angela Fong <angela.fong@joyent.com>
Date:   2017-10-24T18:43:47-07:00 (1 year, 11 months ago)
    
    ADMINUI-2388 When packages with similar names are created, their package versions are listed in the other's versions
    ADMINUI-2392 Package search by memory and disk was broken by ADMINUI-2378
    Reviewed by: Marsell Kukuljevic <marsell@joyent.com>
    Approved by: Marsell Kukuljevic <marsell@joyent.com>

diff --git a/lib/adminui.js b/lib/adminui.js
index d0c47d6e..8b3fedeb 100644
--- a/lib/adminui.js
+++ b/lib/adminui.js
@@ -40,8 +40,11 @@ function resume(req, res, next) {
 }
 
 function paramsTrimmer(req, res, next) {
-    Object.keys(req.params).forEach(function (paramName) {
-        req.params[paramName] = req.params[paramName].trim();
+    var params = req.params;
+    Object.keys(params).forEach(function (paramName) {
+        if (typeof (params[paramName]) === 'string') {
+            params[paramName] = params[paramName].trim();
+        }
     });
     next();
 }
diff --git a/lib/packages.js b/lib/packages.js
index 66da16ca..6b2fcc39 100644
--- a/lib/packages.js
+++ b/lib/packages.js
@@ -126,15 +126,24 @@ module.exports.list = function (req, res, next) {
             if (key === 'group' || key === 'traits') {
                 return;
             }
+            // We need to distinguish the exact name search from partial name search by tweaking the
+            // argument of the latter to 'name_substring'. Exact name search is done when displaying the
+            // different versions of the same package name on the details page. Partial name search
+            // is used on the list page. Both result in the same search key 'name' in papi but have
+            // different argument/filter syntax.
             if (typeof (value) === 'string' && value.length) {
-                if (key === 'billing_id' || value === 'true' || value === 'false') {
+                if (key === 'billing_id' || key === 'name' || value === 'true' || value === 'false') {
                     if (key === 'billing_id') {
                         filter.push(sprintf('(%s=%s)', 'uuid', value));
                     } else {
                         filter.push(sprintf('(%s=%s)', key, value));
                     }
                 } else {
-                    filter.push(sprintf('(%s:caseIgnoreSubstringsMatch:=*%s*)', key, value));
+                    if (key === 'name_substring') {
+                        filter.push(sprintf('(%s:caseIgnoreSubstringsMatch:=*%s*)', 'name', value));
+                    } else {
+                        filter.push(sprintf('(%s:caseIgnoreSubstringsMatch:=*%s*)', key, value));
+                    }
                 }
                 return;
             }
diff --git a/www/js/views/packages.js b/www/js/views/packages.js
index caf76183..8799573a 100644
--- a/www/js/views/packages.js
+++ b/www/js/views/packages.js
@@ -222,7 +222,7 @@ var PackagesView = Backbone.Marionette.Layout.extend({
 
         var data = {
             active: getValue('active'),
-            name: getValue('name'),
+            name_substring: getValue('name'),
             traits: getValue('traits'),
             group: getValue('group'),
             billing_id: getValue('billing_id')
